python/ray/data/_internal/logical/operators/read_operator.py

```diff
@@ -0,0 +1,63 @@
+from typing import Any, Dict, Iterator
+
+import ray
+import ray.cloudpickle as cloudpickle
+from ray.data._internal.execution.interfaces import RefBundle, PhysicalOperator
+from ray.data._internal.execution.operators.map_operator import MapOperator
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.logical.interfaces import LogicalOperator
+from ray.data.block import Block, BlockMetadata
+from ray.data.datasource.datasource import Datasource
+
+
+class Read(LogicalOperator):
+    """Logical operator for read."""
+
+    def __init__(
+        self,
+        datasource: Datasource,
+        parallelism: int = -1,
+        ray_remote_args: Dict[str, Any] = None,
+        read_args: Dict[str, Any] = None,
+    ):
+        super().__init__("Read", [])
+        self._datasource = datasource
+        self._parallelism = parallelism
+        self._ray_remote_args = ray_remote_args
+        self._read_args = read_args
+
+
+def plan_read_op(op: Read) -> PhysicalOperator:
+    """Get the corresponding DAG of physical operators for Read."""
+
+    def get_input_data():
+        reader = op._datasource.create_reader(**op._read_args)
+        read_tasks = reader.get_read_tasks(op._parallelism)
+        return [
+            RefBundle(
+                [
+                    (
+                        # TODO(chengsu): figure out a better way to pass read
+                        # tasks other than ray.put().
```

Comment:
It doesn't need to be in this PR, but I still think that we should have something akin to a `ScanOperator` that subclasses `MapOperator` so we can do away with this `InputDataBuffer` instance that only serves as a hack to feed read tasks to a downstream `MapOperator`.

This would have a bunch of other advantages, like making physical plan optimizations easier to implement such as projection pushdown on a partitioned datasource, and providing a way to do progressive resolving of the read operator.

Reply:
Yeah I originally want to push for a `Scan/ReadOperator` as well. To me, it looks like `Executor` needs to add input to `Scan/ReadOperator` to trigger execution right? I thought we don't want `Executor` to do that? cc @ericl

Reply:
It can be a logical operator, but from the physical execution pov there's no such thing as a scan.